src/array_utils.h

```diff
@@ -56,4 +62,106 @@ shape_from_tuple(nb::tuple new_shape, size_t elem_count)
+)
+{
+    std::vector<int> cpp_axes {};
+    if (std::holds_alternative<nanobind::tuple>(axes)) {
+        for (const auto& element : std::get<nanobind::tuple>(axes)) {
+            cpp_axes.push_back(nanobind::cast<int>(element));
+        }
+
+    } else {
+        cpp_axes.push_back(static_cast<int>(std::get<nanobind::int_>(axes)));
+    }
+
+    std::unordered_set<std::size_t> unique_axes;
+
+    std::vector<std::size_t> result {};
+    if (cpp_axes.size() != n_dim) {
+        throw nb::value_error("Axis don't match array");
+    }
+    for (std::size_t i = 0; i < cpp_axes.size(); i++) {
+
+        int axis = cpp_axes[i];
+        if (axis < 0) {
+            axis = n_dim + axis;
+        }
+        std::size_t pos_axis = static_cast<std::size_t>(axis);
+        if (pos_axis >= n_dim) {
+            throw nb::value_error("Axis don't match array");
+        }
+        if (!unique_axes.insert(pos_axis).second) {
+            throw nb::value_error("Repeated axis in provided axes");
+        }
+        result.push_back(pos_axis);
+    }
+
+    return result;
+}
+
+/**
+ * @brief Permutes the dimensions of the input data and copies it to the destination.
+ *
+ * This function permutes the dimensions of the source data according to the provided
+ * permutation array and copies the permuted data to the destination.
+ * Assumunes that the sizes specified in src_shape and perm match the dimensionality of
+ * the data.
+ */
+template <typename RANDOM_ACCESS_CONST_ITERATOR, typename RANDOM_ACCESS_ITERATOR>
+static APY_INLINE void transpose_axes_and_copy_data(
+    RANDOM_ACCESS_CONST_ITERATOR src,
+    RANDOM_ACCESS_ITERATOR dst,
+    const std::vector<std::size_t>& src_shape,
+    const std::vector<std::size_t>& new_axis,
+    std::size_t itemsize = 1
+)
+{
+    // Calculate the destination shape based on the permutation
+    std::vector<std::size_t> dst_shape(src_shape.size());
+    for (std::size_t i = 0; i < new_axis.size(); ++i) {
+        dst_shape[i] = src_shape[new_axis[i]];
+    }
+
+    std::vector<std::size_t> dst_stride = strides_from_shape(dst_shape);
+
+    // Start at indice [0,0, ..., 0]
+    std::vector<std::size_t> indices(src_shape.size(), 0);
+    std::vector<std::size_t> dst_indices(indices.size());
+
+    std::size_t num_elem = fold_shape(src_shape);
+
+    for (std::size_t src_index = 0; src_index < num_elem; ++src_index) {
+
+        // For each dimension in the destination, assign the corresponding index from
+        // the source based on the new axis. For example, if new_axis = {1, 2, 0}, and
+        // current_indices = {i, j, k}, the resulting dst_indices will be {j, k, i}
+        for (std::size_t j = 0; j < new_axis.size(); ++j) {
+            dst_indices[j] = indices[new_axis[j]];
+        }
+        // Calculate the flat index for destination
+        std::size_t dst_index = 0;
+        for (std::size_t i = 0; i < indices.size(); ++i) {
+            dst_index += dst_indices[i] * dst_stride[i];
```

Comment:
```suggestion
            dst_index += indices[new_axis[i]] * dst_stride[i];
```
? 

But the current code is undoubtedly easier to follow...